pkg/controller/common/association/conf.go

```diff
@@ -0,0 +1,135 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package association
+
+import (
+	"encoding/json"
+	"reflect"
+	"unsafe"
+
+	commonv1alpha1 "github.com/elastic/cloud-on-k8s/pkg/apis/common/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/annotation"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// FetchWithAssociation retrieves an object and extracts its association configuration.
+func FetchWithAssociation(client k8s.Client, request reconcile.Request, obj commonv1alpha1.Associator) (bool, error) {
+	if err := client.Get(request.NamespacedName, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	assocConf, err := GetAssociationConf(obj)
+	if err != nil {
+		return false, err
+	}
+
+	obj.SetAssociationConf(assocConf)
+	return true, nil
+}
+
+// GetAssociationConf extracts the association configuration from the given object by reading the annotations.
+func GetAssociationConf(obj runtime.Object) (*commonv1alpha1.AssociationConf, error) {
+	accessor := meta.NewAccessor()
+	annotations, err := accessor.Annotations(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	return extractAssociationConf(annotations)
+}
+
+func extractAssociationConf(annotations map[string]string) (*commonv1alpha1.AssociationConf, error) {
+	if len(annotations) == 0 {
+		return nil, nil
+	}
+
+	var assocConf commonv1alpha1.AssociationConf
+	serializedConf, exists := annotations[annotation.AssociationConfAnnotation]
+	if !exists || serializedConf == "" {
+		return nil, nil
+	}
+
+	if err := json.Unmarshal(unsafeStringToBytes(serializedConf), &assocConf); err != nil {
+		return nil, errors.Wrapf(err, "failed to extract association configuration")
+	}
```

Comment:
I started by having separate annotations for each field but the implementation started to get a bit too repetitive. Generally, all of the fields contained here are accessed during a reconciliation run (more than once for some fields) so in terms of the lines of code involved to extract them and the associated performance implications, individual annotations per field didn't seem like an attractive option to me. 
- There's no case for "lazy" access so might as well serialize/deserialize a single large blob instead of multiple smaller blobs.
- Increased cognitive load of having to consider how each field should be stored and retrieved (should `AuthSecretName` and `AuthSecretKey` be stored under a single annotation or two, what is the format etc.) 
- Right now, all values are primitive but in the future we may want to include more complex ones (Eg. multiple hosts for APM output) -- which would require some form of serialization anyway.
- Having all the related values stored under a single annotation signals their relationship more strongly than having separate annotations.
- JSON encoded annotation values are not unusual in the K8S world (kubectl, some ingresses, etc.) :smile:

Reply:
👍 except for 

> Right now, all values are primitive but in the future we may want to include more complex ones (Eg. multiple hosts for APM output) -- which would require some form of serialization anyway.

I hope we don't add more attributes to this.